docs/stdlib/fts.rst

```diff
@@ -0,0 +1,575 @@
+.. _ref_std_fts:
+
+.. versionadded:: 4.0
+
+================
+Full-text Search
+================
+
+The ``fts`` built-in module contains various tools that enable full-text
+search functionality in EdgeDB.
+
+.. list-table::
+    :class: funcoptable
+
+    * - :eql:type:`fts::Language`
+      - Common languages :eql:type:`enum`
+
+    * - :eql:type:`fts::PGLanguage`
+      - Postgres languages :eql:type:`enum`
+
+    * - :eql:type:`fts::ElasticLanguage`
+      - ElasticSearch languages :eql:type:`enum`
+
+    * - :eql:type:`fts::LuceneLanguage`
+      - Apache Lucene languages :eql:type:`enum`
+
+    * - :eql:type:`fts::Weight`
+      - Weight category :eql:type:`enum`
+
+    * - :eql:type:`fts::document`
+      - Opaque document type
+
+    * - :eql:func:`fts::search`
+      - :eql:func-desc:`fts::search`
+
+    * - :eql:func:`fts::with_options`
+      - :eql:func-desc:`fts::with_options`
+
+When considering FTS functionality our goal was to come up with an interface
+that could support different backend FTS providers. To achieve that we've
+identified the following components to the FTS functionality:
+
+1) Valid FTS targets must be indexed.
+2) The expected language should be specified at the time of creating an index.
+3) It should be possible to mark document parts as having different relevance.
+4) It should be possible to assign custom weights at runtime so as to make
+   searching more flexible.
+5) The search query should be close to what people are already used to.
+
+To address (1) we introduce a special ``fts::index``. The presence of this
+index in a type declaration indicates that the type in question can be subject
+to full-text search. This is an unusual index as it actually affects the
+results of :eql:func:`fts::search` function. This is unlike most indexes which
+only affect the performance and not the actual results. Another special
+feature of ``fts::index`` is that at most one such index can be declared per
+type. If a type inherits this index from a parent and also declares its own,
+only the new index applies and fully overrides the ``fts::index`` inherited
+from the parent type. This means that when dealing with a hierarchy of
+full-text-searchable types, each type can customize what gets searched as
+needed.
+
```

Comment:
Should we also mention what happens when two parents both define their own fts::index?

The error message is clear enough, so it might not be needed.

https://github.com/edgedb/edgedb/blob/master/edb/schema/indexes.py#L1222-L1230